pandas/core/groupby/groupby.py

```diff
@@ -1670,6 +1671,72 @@ def nth(self, n, dropna=None):
 
         return result
 
+    def quantile(self, q=0.5, interpolation='linear'):
+        """
+        Return group values at the given quantile, a la numpy.percentile.
+
+        Parameters
+        ----------
+        q : float or array-like, default 0.5 (50% quantile)
+            0 <= q <= 1, the quantile(s) to compute
+        interpolation : {'linear', 'lower', 'higher', 'midpoint', 'nearest'}
+            Method to use when the desired quantile falls between two points.
+
+        Returns
+        -------
+        Series or DataFrame
+            Return type determined by caller of GroupBy object.
+
+        See Also
+        --------
+        Series.quantile : Similar method for Series
+        DataFrame.quantile : Similar method for DataFrame
+        numpy.percentile : NumPy method to compute qth percentile
+
+        Examples
+        --------
+        >>> df = pd.DataFrame(
+        ...    [['foo'] * 5 + ['bar'] * 5,
+        ...     [1, 2, 3, 4, 5, 5, 4, 3, 2, 1]],
+        ...    columns=['key', 'val'])
+        >>> df
+        """
+
+        inferences = {  # TODO (py27): replace with nonlocal
+            'is_dt': False,
+            'is_int': False
+        }
+
+        def pre_processor(vals):
+            if vals.dtype == np.object:
+                raise TypeError("'quantile' cannot be performed against "
+                                "'object' dtypes!")
+            elif is_integer_dtype(vals):
+                inferences['is_int'] = True
+            elif is_datetime64_dtype(vals):
+                vals = vals.astype(np.float)
+                inferences['is_dt'] = True
+
+            return vals
+
+        def post_processor(vals):
+            if inferences['is_dt']:
+                vals = vals.astype('datetime64[ns]')
+            elif inferences['is_int'] and interpolation in [
+                    'lower', 'higher', 'nearest']:
+                vals = vals.astype(np.int64)
```

Comment:
@chris-b1 you were spot on with the issue - thanks! Is this the idiomatic way of casting back? Saw it elsewhere in the code base just not sure what implications may carry for windows